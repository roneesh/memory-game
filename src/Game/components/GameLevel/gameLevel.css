.game-board {
  background-color: #4f5d75;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  padding: 3px 0;
}

.game-board__tile {
  width: 100px;
  height: 100px;
  margin: 3px;
  perspective: 300px;
}

.game-board__tile--static {
  perspective: 0;
}

.game-board__tile__card {
  height: 100%;
  width: 100%;
  position: relative;
  transition: transform 1s;
  transform-style: preserve-3d;
  color: #ffffff;
  font-size: 32px;
  font-family: 'Monoton', cursive;
}

.game-board__tile__card__face {
  position: absolute;
  height: 100%;
  width: 100%;
  backface-visibility: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
}

.game-board__tile__card__face--front {
  background-color: #bfc0c0;
}

.game-board__tile__card__face--back {
  background-color: transparent;
}