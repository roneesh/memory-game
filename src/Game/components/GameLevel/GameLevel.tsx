import { shuffle } from 'lodash';
import * as React from 'react';
import './gameLevel.css';
import { LevelPreview, LevelReady, } from './states';

interface IGameLevelProps {
  updateLevel: () => void;
};

type LevelState = "ready" | "preview" | "playing" | "finished";
export interface IBoardTile {
  character: string | HTMLElement;
  visible: boolean;
};
export interface ILevelTurns {
  count: number;
  currentTurn: number[];
}
interface IGameLevelState {
  levelState: LevelState;
  board: IBoardTile[];
  turns:  ILevelTurns;
};

class GameLevel extends React.Component<IGameLevelProps, IGameLevelState> {
  constructor(props: IGameLevelProps) {
    super(props);

    this.state = {
      board: this.generateBoard(),
      levelState: 'ready',
      turns: {
        count: 0,
        currentTurn: [],
      },
    };
  }

  public generateBoard(): IBoardTile[] {
    return shuffle(
      ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
      'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'].map(character => {
        return {
          character,
          visible: false,
        };
      })
    );
  }

  public setLevelState = (levelState: LevelState) => {
    this.setState({ levelState });
  }

  public getResetBtn() {
    const { levelState } = this.state;

    const text = levelState === 'ready' ? 'Play Level' : 'Reset';
    const nextState = levelState === 'ready' ? 'preview' : 'ready';
    return <button onClick={this.setLevelState.bind(null, nextState)}>
      {text}
    </button>;
  }

  public renderControls() {

    return <div className="game-level__controls">
      {this.getResetBtn()}
    </div>
  }

  public renderState() {
    const { levelState, board } = this.state;
    switch(levelState) {
      case 'ready':
        return <LevelReady board={board} />;
      case 'preview':
        return <LevelPreview board={board} />;
      // case 'playing':
      //   return <LevelPlaying />;
      // case 'finished':
      //   return <LevelFinished />;
      default:
        return null;
    }
  }

  public render() {
    return <div className="game-level">
      {this.renderControls()}
      {this.renderState()}
    </div>;
  }
}

export default GameLevel;
