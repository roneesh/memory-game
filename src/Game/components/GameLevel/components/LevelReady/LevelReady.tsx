import * as React from 'react';
import { IBoardTile } from '../../GameLevel';
import './levelReady.css';

interface ILevelReadyProps {
  board: IBoardTile[];
};

const LevelReady = (props: ILevelReadyProps) => {
  const { board } = props;

  return <div className="game-board">
    {board.map((item, index) => {
      return <div className="game-board__tile" key={index} />;
    })}
  </div>
};

export default LevelReady;
