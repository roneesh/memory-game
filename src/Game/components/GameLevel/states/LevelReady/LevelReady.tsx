import * as React from 'react';
import { IBoardTile } from '../../GameLevel';

interface ILevelReadyProps {
  board: IBoardTile[];
};

const LevelReady = (props: ILevelReadyProps) => {
  const { board } = props;

  return <div className="game-board">
    {board.map((item, index) => {
      return <div className="game-board__tile game-board__tile--static" key={index}>
        <div className="game-board__tile__card__face game-board__tile__card">
          <div className="game-board__tile__card__face game-board__tile__card__face--front" />
        </div>
      </div>;
    })}
  </div>
};

export default LevelReady;
