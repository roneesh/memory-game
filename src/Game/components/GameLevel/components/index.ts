import LevelReady from './LevelReady';
// import LevelPlaying from './LevelPlaying';
// import LevelFinished from './LevelFinished';
// import LevelPreview from './LevelPreview';

export {
  LevelReady,
  // LevelPlaying,
  // LevelFinished,
  // LevelPreview,
};
