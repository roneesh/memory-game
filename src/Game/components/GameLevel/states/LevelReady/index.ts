import LevelReady from './LevelReady';

export default LevelReady;
