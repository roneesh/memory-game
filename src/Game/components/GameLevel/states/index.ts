// import LevelFinished from './LevelFinished';
// import LevelPlaying from './LevelPlaying';
import LevelPreview from './LevelPreview';
import LevelReady from './LevelReady';

export {
  // LevelFinished,
  // LevelPlaying,
  LevelPreview,
  LevelReady,
};
