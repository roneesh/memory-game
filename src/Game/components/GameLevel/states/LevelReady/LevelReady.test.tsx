import * as React from 'react';
import renderer from 'react-test-renderer';
import { IBoardTile } from '../../GameLevel';
import LevelReady from './LevelReady';

describe('The level when its ready', () => {
  let board: IBoardTile[] = [];
  const boardBase =
    ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
    'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'];

  beforeEach(() => {
    board = boardBase.map(character => {
      return {
        character,
      };
    });
  });

  it('it renders', () => {
    const levelReady = renderer
      .create(<LevelReady board={board} />)
      .toJSON();
    expect(levelReady).toMatchSnapshot();
  });
});
