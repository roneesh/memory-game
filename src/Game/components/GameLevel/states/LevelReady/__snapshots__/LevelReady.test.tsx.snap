// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The level when its ready it renders 1`] = `
<div
  className="game-board"
>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
  <div
    className="game-board__tile game-board__tile--static"
  >
    <div
      className="game-board__tile__card__face game-board__tile__card"
    >
      <div
        className="game-board__tile__card__face game-board__tile__card__face--front"
      />
    </div>
  </div>
</div>
`;
